packages/cohort_banking/src/callbacks/statemap_installer.rs

```diff
@@ -1,68 +1,106 @@
+        let updated_rows = tx.execute(BANK_ACCOUNTS_UPDATE_QUERY, params).await.map_err(|e| e.to_string())?;
+        Ok(updated_rows)
+    }
 
-                let sql = r#"
-                    UPDATE bank_accounts ba SET
-                        "amount" =
-                            (CASE
-                                WHEN ba."number" = ($1)::TEXT THEN ba."amount" + ($3)::DECIMAL
-                                WHEN ba."number" = ($2)::TEXT THEN ba."amount" - ($3)::DECIMAL
-                            END),
-                        "version" = ($4)::BIGINT
-                    WHERE ba."number" IN (($1)::TEXT, ($2)::TEXT)
-                        AND ba."version" < ($4)::BIGINT
-                    "#;
+    async fn update_snapshot(tx: &Transaction<'_>, snapshot_version: u64) -> Result<u64, String> {
+        let params: &[&(dyn ToSql + Sync)] = &[&(snapshot_version as i64), &"SINGLETON"];
 
-                let params: &[&(dyn ToSql + Sync)] = &[&request.from, &request.to, &request.amount, &(sti.version as i64)];
+        tx.execute(SNAPSHOT_UPDATE_QUERY, params).await.map_err(|e| e.to_string())
+    }
+}
 
-                let updated_rows = tx.execute(sql, params).await.map_err(|e| e.to_string())?;
+#[async_trait]
+impl ReplicatorInstaller for BankStatemapInstaller {
+    /// Install statemaps and version for respective rows in the `bank_accounts` table and update the `snapshot_version` in `cohort_snapshot` table.
+    ///
+    /// Certain errors like `could not serialize access due to concurrent update` in postgres are retriable.
+    /// - Updating `bank_account` workflow.
+    ///     - On successful completion, we proceed to update the `snapshot_version` in `cohort_snapshot`.
+    ///     - When there is a retryable error, go to the start of the loop to retry. `(Txn aborts and retries)`
+    ///     - For non-retryable error, we return the `Error`.
+    /// - Updating `cohort_snapshot` workflow.
+    ///     - On successful completion, we commit the transaction and return. `(Txn Commits and returns)`
+    ///     - When there is a retryable error, go to the start of the loop to retry. `(Txn aborts and retries)`
+    ///     - For non-retryable error, we return the `Error`.
+
+    async fn install(&self, statemap: Vec<StatemapItem>, snapshot_version: u64) -> Result<(), String> {
+        let mut cnn = self.database.get().await.map_err(|e| e.to_string())?;
+        loop {
+            let tx = cnn.transaction().await.map_err(|e| e.to_string())?;
+            if !statemap.is_empty() {
+                let updated_rows_res = BankStatemapInstaller::install_bank_transfer_statemap(&tx, &statemap, snapshot_version).await;
+
+                match updated_rows_res {
+                    Ok(updated_rows) => {
+                        if updated_rows > 0 {
+                            log::debug!(
+                                "No rows were updated when installing: {:?}. Snapshot will be set to: {}",
```

Comment:
Is this typo? Says "No rows"

Reply:
Aah I never paid attention to this. Good catch. I just moved around the original if block that you had written. I will update this.

![image](https://github.com/kindredgroup/talos/assets/118979108/6c1b5aae-20f6-4616-9da7-c782fc5d8163)